src/core/ucc_mc.c

```diff
@@ -86,40 +96,22 @@ ucc_status_t ucc_mc_type(const void *ptr, ucc_memory_type_t *mem_type)
 
 ucc_status_t ucc_mc_alloc(void **ptr, size_t size, ucc_memory_type_t mem_type)
 {
-    ucc_status_t status;
-
-    if (NULL == mc_ops[mem_type]) {
-        ucc_error("no components supported memory type %s available",
-                  ucc_memory_type_names[mem_type]);
-        return UCC_ERR_NOT_SUPPORTED;
-    }
-    status = mc_ops[mem_type]->mem_alloc(ptr, size);
-
-    return status;
+    UCC_CHECK_MC_AVAILABLE(mem_type);
+    return mc_ops[mem_type]->mem_alloc(ptr, size);
 }
 
-ucc_status_t ucc_mc_reduce(const void *src1, ucc_memory_type_t src1_mt,
-                           const void *src2, ucc_memory_type_t src2_mt,
-                           void *dst, ucc_memory_type_t dst_mt, size_t count,
-                           ucc_datatype_t dt, ucc_reduction_op_t op)
+ucc_status_t ucc_mc_reduce(const void *src1, const void *src2,
+                           void *dst, size_t count, ucc_datatype_t dt,
+                           ucc_memory_type_t mem_type, ucc_reduction_op_t op)
 {
-    //TODO
-
-    return UCC_ERR_NOT_IMPLEMENTED;
+    UCC_CHECK_MC_AVAILABLE(mem_type);
+    return mc_ops[mem_type]->reduce(src1, src2, dst, count, dt, op);
 }
 
 ucc_status_t ucc_mc_free(void *ptr, ucc_memory_type_t mem_type)
 {
-    ucc_status_t status;
-
-    if (NULL == mc_ops[mem_type]) {
-        ucc_error("no components supported memory type %s available",
-                  ucc_memory_type_names[mem_type]);
-        return UCC_ERR_NOT_SUPPORTED;
-    }
-    status = mc_ops[mem_type]->mem_free(ptr);
-
-    return status;
+    UCC_CHECK_MC_AVAILABLE(mem_type);
```

Comment:
do we need the check here? assert maybe is enough. if you call free this means you already succeeded allocating the ptr of this mem_type, so component should be available.